### PR TITLE
[processing] Use native QGIS map layer combobox 

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR adds a bunch of functionality to QgsMapLayerComboBox which was present in processing's custom map layer combobox, and then switches processing across to use the standard map layer combox when you run an algorithm outside the toolbox.

Benefits include:
- reduction of duplicate code, removal of code with no unit testing
- proper sorting of layers
- icons for layer types
- combobox responds immediately to layer addition/removal

### CODE REVIEW

**Comment by nirvn** [comment]
That's great, it'll help _a lot_ in making processing feel and act like the rest of QGIS.

File: src/core/qgsmaplayermodel.h [hunk 1]
```diff
@@ -32,13 +32,21 @@ class QgsMapLayer;
...
+
   public:
 
     //! Item data roles
     enum ItemDataRole
     {
       LayerIdRole = Qt::UserRole + 1, //!< Stores the map layer ID
       LayerRole, //!< Stores pointer to the map layer itself
+      IsEmptyRole, //!< True if index corresponds to the empty (not set) value
+      IsAdditionalRole, //!< True if index corresponds to an additional (non map layer) item
```
Comment (3nids) [inline src/core/qgsmaplayermodel.h]: For consistency, better not having `Is*` ?

File: src/core/qgsmaplayermodel.cpp [hunk 2]
```diff
@@ -35,6 +37,7 @@ QgsMapLayerModel::QgsMapLayerModel( QObject *parent )
     : QAbstractItemModel( parent )
     , mLayersChecked( QMap<QString, Qt::CheckState>() )
     , mItemCheckable( false )
+    , mAllowEmpty( false )
```
Comment (3nids) [inline src/core/qgsmaplayermodel.cpp]: missing `mShowCrs` definition?

**Comment by 3nids** [comment]
Cool, thanks for this!

File: src/core/qgsmaplayermodel.cpp (hunk 2 above)
Comment (nyalldawson) [inline src/core/qgsmaplayermodel.cpp]: Thanks! That's probably why Travis was failing

**Pull request merged**